### PR TITLE
Version 3.1 Release Candidate

### PR DESCRIPTION
**Pull request by chris-cheshire**
### Major Changes

- IgG controls will now be analysed by the deeptools QC subworkflow giving greater visibility on the quality of control samples.
- Updated the MACS2 default parameters to better process PA-Tn5/PA-Mnase based experiments. The new defaults use the q-value of `0.01` as the default cutoff in place of the p-value. The defaults have also been updated to keep duplicate reads int he peak finding process and also to shift the model to better account for nucleosome positioning `--nomodel --shift -75 --extsize 150 --keep-dup all`

### Enhancements

- Updated pipeline template to nf-core/tools `2.7.2`.
- Updated pipeline syntax to conform to new Nextflow version standards.
- Some locally defined subworkflows/modules have now been added to nf-core and re-imported as official modules/subworkflows.

### Fixes

- Fixed confusing config warnings that were being displayed on legitmate parameter configurations.
- Fixed deeptools correlation plots that were showing low levels of correlation even in test data by changing the plot to use the pearson test.
- Corrected the SEACR p-value parameter description.
- Fixed output of picard mark/remove duplicate files so that the sorted, indexed bams for all files always output to the results folder.
- Spikein genome processes and checks no longer run when the normalisation mode is set to something other than `SpikeIn`.

### Software dependencies

Note, since the pipeline is now using Nextflow DSL2, each process will be run with its own [Biocontainer](https://biocontainers.pro/#/registry). This means that on occasion it is entirely possible for the pipeline to be using different versions of the same tool. However, the overall software dependency changes compared to the last release have been listed below for reference.

| Dependency | Old version | New version |
| ---------- | ----------- | ----------- |
| `multiqc`  | 1.14        | 1.16.1      |
| `samtools` | 2.27.2      | 2.27.4      |

> **NB:** Dependency has been **updated** if both old and new version information is present.
> **NB:** Dependency has been **added** if just the new version information is present.
> **NB:** Dependency has been **removed** if version information isn't present.

### CODE REVIEW

File: lib/WorkflowCutandrun.groovy [hunk 1]
```diff
@@ -18,7 +18,7 @@ class WorkflowCutandrun {
             System.exit(1)
         }
 
-        if (!params.spikein_fasta) {
+        if (params.normalisation_mode == "Spikein" && !params.spikein_fasta) {
```
Comment (SPPearce) [inline lib/WorkflowCutandrun.groovy:21]: What happens if `normalisation_mode` is given as "spikein"

File: nextflow.config [hunk 2]
```diff
@@ -101,7 +102,9 @@ params {
     dt_heatmap_peak_beforelen  = 3000
     dt_heatmap_peak_afterlen   = 3000
     dt_qc_bam_binsize          = 500
+    dt_qc_corr_method          = "pearson"
     fragment_size              = 100 // PARAM NOT USED
```
Comment (SPPearce) [inline nextflow.config:106]: Why is this here if not used?

File: nextflow_schema.json [hunk 3]
```diff
@@ -346,7 +346,7 @@
                     "type": "number",
                     "default": 0.05,
                     "fa_icon": "fas fa-align-justify",
-                    "description": "SEACR p-value threshold for peaks"
+                    "description": "SEACR specifies a numeric threshold of n between 0 and 1 returns the top n fraction of peaks based on total signal within peaks. This is only used if there are no controls included with the samples and if `--use_control` is `false`"
```
Comment (SPPearce) [inline nextflow_schema.json:349]:
```suggestion
                    "description": "SEACR specifies returns the top n fraction (between 0 and 1) of peaks based on total signal within peaks. This is only used if there are no controls included with the samples and if `--use_control` is `false`"
```

File: CHANGELOG.md [hunk 4]
```diff
@@ -3,59 +3,90 @@
...
+- Updated pipeline template to nf-core/tools `2.7.2`.
+- Updated pipeline syntax to conform to new Nextflow version standards.
+- Some locally defined subworkflows/modules have now been added to nf-core and re-imported as official modules/subworkflows.
 
-We thank Harshil Patel ([@drpatelh](https://github.com/drpatelh)) and everyone in the Luscombe Lab ([@luslab](https://github.com/luslab)) for their extensive assistance in the development of this pipeline.
+### Fixes
 
-### Pipeline summary
+- Fixed confusing config warnings that were being displayed on legitmate parameter configurations.
+- Fixed deeptools correlation plots that were showing low levels of correlation even in test data by changing the plot to use the pearson test.
```
Comment (SPPearce) [inline CHANGELOG.md:23]:
```suggestion
- Fixed deeptools correlation plots that were showing low levels of correlation even in test data by changing the plot to use Pearson correlation.
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -3,59 +3,90 @@
...
+- Corrected the SEACR p-value parameter description.
+- Fixed output of picard mark/remove duplicate files so that the sorted, indexed bams for all files always output to the results folder.
```
Comment (SPPearce) [inline CHANGELOG.md:25]:
```suggestion
- Fixed output of Picard mark/remove duplicate files so that the sorted, indexed bams for all files always output to the results folder.
```

File: CHANGELOG.md [hunk 6]
```diff
@@ -3,59 +3,90 @@
...
+### Software dependencies
 
-## [1.1] - 2022-01-05
+Note, since the pipeline is now using Nextflow DSL2, each process will be run with its own [Biocontainer](https://biocontainers.pro/#/registry). This means that on occasion it is entirely possible for the pipeline to be using different versions of the same tool. However, the overall software dependency changes compared to the last release have been listed below for reference.
 
-### Enhancements & fixes
+| Dependency | Old version | New version |
+| ---------- | ----------- | ----------- |
+| `multiqc`  | 1.14        | 1.16.1      |
+| `samtools` | 2.27.2      | 2.27.4      |
```
Comment (SPPearce) [inline CHANGELOG.md:36]: samtools version is 1.16.1 from 1.15.1, should the top be samtools? 
Not sure what the second line might be, if anything.

**Review by SPPearce — APPROVED** [review]
Note: this is my first pipeline code review.

I've gone through and made a few comments, nothing major though.

**Review by DSchreyer — APPROVED** [review]
Everything looks good to me. I made a few comments about the modules, but they might be not relevant. The documentation is very clean and I had no issue understanding the modules and updates. Well done.

 I currently run the pipeline with the test profile on our server, but I don't see an issue why it should not go through. I let you know if I notice something, but otherwise, I approve all commits.

File: lib/WorkflowCutandrun.groovy (hunk 1 above)
Comment (chris-cheshire) [reply lib/WorkflowCutandrun.groovy:21]: It will error but gracefully saying it doesn't recognise the param

```
ERROR: Validation of pipeline parameters failed!


* --normalisation_mode: 'TEST' is not a valid choice (Available choices: Spikein, RPKM, CPM, BPM, None)
```

**Comment by chris-cheshire** [comment]
> 

thank you so much!

File: nextflow.config (hunk 2 above)
Comment (chris-cheshire) [reply nextflow.config:106]: it is used in this code section

```
DEEPTOOLS_QC (
                ch_samtools_bam,
                ch_samtools_bai,
                params.dt_qc_corr_method
)
```

File: nextflow_schema.json (hunk 3 above)
Comment (chris-cheshire) [reply nextflow_schema.json:349]: Updated thanks

File: CHANGELOG.md (hunk 4 above)
Comment (chris-cheshire) [reply CHANGELOG.md:23]: Updated

File: CHANGELOG.md (hunk 5 above)
Comment (chris-cheshire) [reply CHANGELOG.md:25]: Updated

File: CHANGELOG.md (hunk 6 above)
Comment (chris-cheshire) [reply CHANGELOG.md:36]: good spot thanks

File: modules/nf-core/bedtools/complement/main.nf [hunk 7]
```diff
@@ -21,6 +21,7 @@ process BEDTOOLS_COMPLEMENT {
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
+    if ("$bed" == "${prefix}.bed") error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
```
[the remark this answers is not in the document]
Comment (chris-cheshire) [reply modules/nf-core/bedtools/complement/main.nf:24]: Yes although this is a standard feature across all nf-core modules with the risk of a name collision. This would best be raised as an issue on `nf-core/modules` as we just update from there.

File: modules/nf-core/bedtools/intersect/main.nf [hunk 8]
```diff
@@ -21,6 +21,9 @@ process BEDTOOLS_INTERSECT {
     script:
     def args = task.ext.args ?: ''
     def prefix = task.ext.prefix ?: "${meta.id}"
+    if ("$intervals1" == "${prefix}.${extension}" ||
+        "$intervals2" == "${prefix}.${extension}")
+        error "Input and output names are the same, use \"task.ext.prefix\" to disambiguate!"
```
[the remark this answers is not in the document]
Comment (chris-cheshire) [reply modules/nf-core/bedtools/intersect/main.nf:26]: ah ok! haha

File: lib/WorkflowCutandrun.groovy (hunk 1 above)
Comment (SPPearce) [reply lib/WorkflowCutandrun.groovy:21]: Ok, that's fine!
